### DISCUSSION TITLE
Data source: web scraping 🚫

### DISCUSSION THREAD

**Discussion by raphodn** [openfoodfacts/open-prices#311] [Ideas]
One source of data could be web scraping.

### Allowed ? NO

:warning: **Currently we don't allow this as a source** :warning: 
- because of legal reasons
- we focus on prices in physical shops

### Additional info

In the [about page](https://prices.openfoodfacts.org/about) & in the [OFF support](https://support.openfoodfacts.org/help/en-gb/29-open-prices/98-do-you-consider-scraping-prices-from-retailers-websites)
> For legal and technical reasons, we don't consider scraping prices from retailers' websites as a valid way to contribute to Open Prices. We want to make sure that the prices we collect are accurate and up-to-date, and receiving scraped prices from contributors doesn't allow us to do that.
> Price scraping is a considered option in a future version of Open Prices, but it would be done by Open Prices itself so that we can have a proof of the price based on the HTML page.

### Related issues

- https://github.com/openfoodfacts/open-prices/issues/512

**Comment by raphodn** [comment]
But if this were to change, here is a list of shops where data (product EAN, price) seems to be available online : 

|Shop|Flavor(s)|Link|EAN ? (URL, web page, code)|Comments|
|---|---|---|---|---|
|Intermarché (France)|OxF|[3700214618240](https://www.intermarche.com/produit/pate-a-tartiner-bio/3700214618240)|- in the URL<br>- in the code||
|Chronodrive (France)|OxF|[3502110004239](https://www.chronodrive.com/tropicana--pure-premium---pur-jus-de-pamplemousse-rose-presse-P104372)|- in the web page||
|Biocoop (France)|OxF|[3760076003417](https://www.biocoop.fr/praline-noisette-50-pralinut-300g-np0047-000.html)|- in the code (image URL)||
|Naturalia (France)|OxF|[3760076004698](https://www.naturalia.fr/produit/pate-choko-ssucre-190g)|- in the code (image URL)||
|Satoriz (France)|OxF|[5411788026253](https://satoriz-comboire.bio/products/pu789186)|- in the web page||
|Sherpa (France)||[7613031423958](https://www.sherpa.net/supermarche/mijoux/fr/cafes-thes-infusions-poudres-chocolatees/187-chicoree-et-cafe-solubles-7613031423958.html)|- in the URL<br>- in the code|- prestashop|
|Mon Epicerie Paris (France)|OFF|[4020943139359](https://www.monepicerieparis.fr/condiments-epices-huiles/403-no-fish-sauce-arche-4020943139359.html)|- in the URL<br>- in the code||
|Picard (France)|OFF|[3270160168125](https://www.picard.fr/produits/4-escalopes-dinde-viennoise-000000000000016812.html)|- in the code (`gtin13`)||
|La Fourche (France)|OxF|[3760324840139](https://lafourche.fr/products/la-fourche-pesto-vert-bio-0-18kg-ht)|- in the code||
|O'zando (France)|OFF|[8720812311584](https://ozando.fr/yellow-plantain-fufu-african-beauty-681g-3/)|- in the web page||
||||||
|Castorama|OPF|[3086960203702](https://www.castorama.fr/pot-de-fleurs-eda-plastiques-coloris-taupe-en-plastique-30-x-h-25-5-cm-contenance-10-l/3086960203702_CAFR.prd)|- in the URL<br>- in the web page||
|Leroy Merlin|OPF||- in the code ?||
|Botanic|OxF|[3086960225285](https://www.botanic.com/produit/231168/pot-blanc-eda-toscane-o-80-x-h-66-cm.html)|- in the code||
||||||
|Joué Club|OPF|[8412668200074](https://www.joueclub.fr/jeux-de-societe/100-logique-8412668200074.html)|- in the URL<br>- in the web page||
|King Jouet|OPF|[0194735331857](https://www.king-jouet.com/jeu-jouet/jeux-societes/jeux-cartes/ref-997021-uno-zero.htm)|- in the web page||
|Smart One|OPF|[8053634003222](https://www.smart-one.fr/lampes-ampoules-deco/4766-itotal-lampe-paillette-rocket-noir-lampes-et-ampoules-8053634003222.html)|- in the URL<br>- in the web page||
||||||
|LaSante.net|OBF, OPF|[3551101034366](https://lasante.net/espace-para/bien-etre-et-minceur/bien-etre-complements-alimentaires/delical-concentre-boisson-hphc-sans-lactose-saveur-vanille-4-x-200-ml.html)|- in the web page||

- https://www.orientalpanda.com/fr/

**Reply by raphodn** [reply]
https://github.com/badlogic/heissepreise
- https://shop.rewe.de/api/products/?search=4002846034603
- https://www.aldi-nord.de/ (seems to not provide GTINs)

**Comment by CloCkWeRX** [comment]
alltheplaces.xyz does a lot of scraping where business have published schema.org data of their location.

The equivalent would be 

- https://schema.org/Offer (has price)
- https://schema.org/Product

Google encourages folks to do this:
https://developers.google.com/search/docs/specialty/ecommerce/include-structured-data-relevant-to-ecommerce

I would guess a significant chunk of supermarkets publish some price data like this.

**Reply by CloCkWeRX** [reply]
Example of one of your URLs:

https://validator.schema.org/#url=https%3A%2F%2Fwww.botanic.com%2Fproduit%2F231168%2Fpot-blanc-eda-toscane-o-80-x-h-66-cm.html

![image](https://github.com/user-attachments/assets/154aa0b3-e8bc-443a-b61f-df47045889de)
![image](https://github.com/user-attachments/assets/b69b44c2-49b6-4a2f-b2d0-64eb600bee9f)

**Reply by CloCkWeRX** [reply]
As a proof of concept, I created https://github.com/CloCkWeRX/alltheprices based on alltheplaces.xyz.

Example output:
```
2025-04-06 03:21:20 [scrapy.core.engine] DEBUG: Crawled (200) <GET https://www.botanic.com/produit/448109/hibiscus.html> (referer: https://www.botanic.com/Assets/Rbs/Seo/100541/fr_FR/Rbs_Catalog_Product.1.xml)
2025-04-06 03:21:20 [scrapy.core.scraper] DEBUG: Scraped from <200 https://www.botanic.com/produit/448109/hibiscus.html>
{'extras': {'@source_uri': 'https://www.botanic.com/produit/448109/hibiscus.html'},
 'image': 'https://botanic-botanic-storage.omn.proximis.com/Imagestorage/imagesSynchro/0/0/ea984d27164cea9bf70e210c6316a365b823d761_1536742.png',
 'name': 'HIBISCUS',
 'offers': {'@type': 'Offer',
            'availability': 'https://schema.org/InStock',
            'price': 23.99,
            'priceCurrency': 'EUR',
            'url': 'https://www.botanic.com/produit/448109/hibiscus.html'},
 'ref': '448109',
 'website': 'https://www.botanic.com/produit/448109/hibiscus.html'}
```

I will think about the best identifiers (gtin) to require, write a few more, think about filtering or mapping to openfoodfacts equivalent categories

**Reply by CloCkWeRX** [reply]
Did the first crawl via https://github.com/CloCkWeRX/alltheprices/releases/tag/2025-04-13

Unfortunately, GTINs are few and far between. Need to do a bit of work for normalisation of naming, extracting other attributes, etc. Need to think about offers structure(s)

**Reply by raphodn** [reply]
Hi, thanks for the investigation! 
I didn't know about schema.org :eyes: 
In any case, we won't be able to upload them anytime soon... :(

**Reply by CloCkWeRX** [reply]
So, I figure I'm going to make a complementary dataset slowly over time. What would be interesting is seeing where schemas can just be aligned or common services used; so that UIs could display/query both without having combined them in the ODBL sense.

**Comment by leonarf** [comment]
Hello, I have written a script that gather EAN and price per kilogram for 3 type of shop (carrefour, U, and intermarche) through their website. It use https://github.com/ultrafunkamsterdam/nodriver to navigate throught their website, selecting a given shop, and making request in the search bar. If you're interested, I can share it.
